### PR TITLE
Fix crash on TSEmptyBodyFunctionExpression AST node

### LINKED ISSUE #2559: Crash on TSEmptyBodyFunctionExpression AST node
I use [eslint-config-airbnb-typescript](https://www.npmjs.com/package/eslint-config-airbnb-typescript). When linting the file copied below, eslint crash with this error: 

```
Error: TSEmptyBodyFunctionExpression ASTNodes are not handled by markPropTypesAsUsed
Occurred while linting .../a/Thing.d.ts:2
    at markPropTypesAsUsed (.../node_modules/eslint-plugin-react/lib/util/usedPropTypes.js:335:15)
    at ObjectPattern (.../node_modules/eslint-plugin-react/lib/util/usedPropTypes.js:526:9)
    at updatedRuleInstructions.(anonymous function) (.../node_modules/eslint-plugin-react/lib/util/Components.js:886:47)
    at listeners.(anonymous function).forEach.listener (.../node_modules/eslint/lib/linter/safe-emitter.js:45:58)
    at Array.forEach (<anonymous>)
    at Object.emit (.../node_modules/eslint/lib/linter/safe-emitter.js:45:38)
    at NodeEventGenerator.applySelector (.../node_modules/eslint/lib/linter/node-event-generator.js:254:26)
    at NodeEventGenerator.applySelectors (.../node_modules/eslint/lib/linter/node-event-generator.js:283:22)
    at NodeEventGenerator.enterNode (.../node_modules/eslint/lib/linter/node-event-generator.js:297:14)
    at CodePathAnalyzer.enterNode (.../node_modules/eslint/lib/linter/code-path-analysis/code-path-analyzer.js:634:23)
```

File to reproduce:
```
// Thing.d.ts
declare class Thing {
  constructor({ id }: { id: string });
}

export default Thing;
```

eslint version: 6.8.0
eslint-plugin-react version:  7.16.0

### PR DESCRIPTION
**Pull request by guillaumewuip**
Fixes #2559

### CODE REVIEW

**Comment by guillaumewuip** [comment]
Ok, no problem 👍 

I'm not sure about how to test this correctly. As far as I understand the test suite, `usedPropTypesInstructions` is not tested yet.

Should I add a test for `usedPropTypesInstructions` mocking `context`, `components` and `utils` params ?

**Comment by golopot** [comment]
The test should contain the code that make this plugin crash, and written like ones in  [tests/lib/rules/no-unused-prop-types.js](https://github.com/yannickcr/eslint-plugin-react/blob/master/tests/lib/rules/no-unused-prop-types.js).

File: tests/util/usedPropTypes.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+// ------------------------------------------------------------------------------
+// Tests
+// ------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester({parserOptions});
+
+ruleTester.run('no-unused-prop-types', rule, {
+  valid: [
+    {
```
Comment (guillaumewuip) [inline tests/util/usedPropTypes.js]: This valid code fails without the fix
image

File: tests/util/usedPropTypes.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      code: [
+        'declare class Thing {',
+        '  constructor({ id }: { id: string });',
+        '}',
+        'export default Thing;'
+      ].join('\n'),
+      parser: parsers.TYPESCRIPT_ESLINT
+    }
+  ],
+  invalid: [
```
Comment (guillaumewuip) [inline tests/util/usedPropTypes.js]: I'm forced to include an invalid test for `RuleTester` to work

**Comment by guillaumewuip** [comment]
@ljharb @golopot I've added one test. Let me know if it's ok for you

File: tests/util/usedPropTypes.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  }
+};
+
+// ------------------------------------------------------------------------------
+// Tests
+// ------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester({parserOptions});
+
+ruleTester.run('no-unused-prop-types', rule, {
```
Comment (ljharb) [inline tests/util/usedPropTypes.js]: since these are tests using this rule, let's please move them to the test files for that rule.

Comment (guillaumewuip) [reply tests/util/usedPropTypes.js]: Ok, done. It's in `tests/lib/rules/no-unused-prop-types.js` now

**Comment by NathanKleekamp** [comment]
Any idea when a new release with this merged PR will go out?

**Pull request merged**